packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -0,0 +1,120 @@
+}
+
+interface ChainEventEmitterInterface {
+  registerChannel(
+    channelId: Bytes32,
+    assetHolders: Address[],
+    listener: ChainEventSubscriberInterface
+  ): void;
+}
+
+interface ChainModifierInterface {
+  fundChannel(arg: FundChannelArg): Promise<providers.TransactionResponse>;
+}
+
+export class ChainService implements ChainModifierInterface, ChainEventEmitterInterface {
+  private readonly ethWallet: NonceManager;
+  private provider: providers.JsonRpcProvider;
+  private addressToObservable: Map<Address, Observable<FundingChangedArg>> = new Map();
+
+  constructor(provider: string, pk: string, pollingInterval?: number) {
+    this.provider = new providers.JsonRpcProvider(provider);
+    if (pollingInterval) this.provider.pollingInterval = pollingInterval;
+    this.ethWallet = new NonceManager(new Wallet(pk, new providers.JsonRpcProvider(provider)));
+  }
+
+  // Only used for unit tests
+  async destructor(): Promise<void> {
+    this.provider.removeAllListeners();
+    this.provider.polling = false;
+  }
+
+  // todo: only works with eth-asset-holder
+  fundChannel(arg: FundChannelArg): Promise<providers.TransactionResponse> {
+    //todo: add retries
+    const transactionRequest = {
+      ...createETHDepositTransaction(arg.channelId, arg.expectedHeld, arg.amount),
+      to: arg.assetHolderAddress,
+      value: arg.amount,
+    };
+    return this.ethWallet.sendTransaction({
+      ...transactionRequest,
+    });
+  }
+
+  registerChannel(
+    channelId: Bytes32,
+    assetHolders: Address[],
+    subscriber: ChainEventSubscriberInterface
+  ): void {
+    assetHolders.map(async assetHolder => {
+      let obs = this.addressToObservable.get(assetHolder);
+      if (!obs) {
+        obs = this.createContractObservable(assetHolder, channelId, subscriber);
+        this.addressToObservable.set(assetHolder, obs);
+      }
+      obs
+        .pipe(filter(event => event.channelId === channelId))
+        // todo: subscriber method should be based on event type
+        .subscribe({next: subscriber.onFundingChanged});
+    });
+  }
+
+  private createContractObservable(
+    contractAddress: Address,
+    channelId: Bytes32,
+    subscriber: ChainEventSubscriberInterface
+  ): Observable<FundingChangedArg> {
+    const contract: Contract = new Contract(
+      contractAddress,
+      ContractArtifacts.EthAssetHolderArtifact.abi
+    ).connect(this.provider);
+
+    // Create an observable that emits events on contract events
+    const obs = new Observable<FundingChangedArg>(subscriber => {
+      // todo: add other event types
+      contract.on('Deposited', (destination, amountDeposited, destinationHoldings) =>
+        subscriber.next({
+          channelId: destination,
+          assetHolderAddress: contractAddress,
+          amount: BN.from(destinationHoldings),
+        })
+      );
+    });
+
+    // Fetch the current contract holding, and emit as an event
+    contract.holdings(channelId).then((holding: any) => {
```

Comment:
Is this to fetch the current holdings when first subscribing to the observable? If so could we use  `merge`  like we do [here](https://github.com/statechannels/statechannels/blob/348e627ae94c1dac10d14a028fe31434bce9db8c/packages/xstate-wallet/src/chain.ts#L205)? That way the `createContractObservable` method wouldn't need to take in a `subscriber`

Reply:
That's good feedback, and I'll play around with that. I would like to keep `registerChannel` as a synchronous method, so let me see if I can make that work.

Reply:
You could fetch it asynchronously and concat instead of merge if you want to guarantee the order

Reply:
`concat` worked well! b1279f5f2